SPIFlashLogger.class.nut

```diff
@@ -224,6 +224,68 @@ class SPIFlashLogger {
         });
     }
 
+    function readSync(index = 1) {
+        assert(typeof index == "integer" && index != 0);
+
+        local skipped = -math.abs(index) + 1
+
+        local readSector;
+        readSector = function(i) {
+             if (i >= _sectors)  // Read requested index greater than our number of logs - return null;
+                return null;
+
+            // start reading sectors
+            // convert sector index `i`, ordered by recency, to physical `sector`, ordered by position on disk
+            local sector;
+            if (index > 0) {
+                sector = (_at_sec + i + 1) % _sectors;
+            } else {
+                sector = (_at_sec - i + _sectors) % _sectors;
+            }
+
+            local addrs_b = _getObjAddrs(sector);
+
+            if (addrs_b.len() == 0)
+                return readSector(i+1)  //TODO: Get rid of my good friend syncronous recursion?
+
+            if (index < 0) // negative index, go backwards
+                addrs_b.seek(-2, 'e');
+
+            local addr, spi_addr, obj, readObj, seekTo;
+
+            if (index < 0) seekTo = -2;
+            else seekTo = 2
+
+            while(++skipped != 1) {  //TODO: Get rid of my good friend the while loop?
+                if ((addrs_b.seek(seekTo, 'c') == -1 || addrs_b.eos() == 1)) {
+                     //  ^ Try to seek to the next available object
+                     // If we've exhausted all addresses found in this sector, move on to the next
+                     return readSector(i + 1) //TODO: Get rid of my good friend syncronous recursion?
```

Comment:
Aron is right - the larger the index, the more hateful the recursion.  This probably should be refactored to use some more conventional loop structures...

For my purposes, I almost always will call *read(1)* if I am doing things sync, and will usethe async methods otherwise so this hasn't been an issue for me.